shopelectro/tasks.py

```diff
@@ -61,18 +63,30 @@ def update_catalog():
         collect_static()
     ]
 
-# @todo #690:60m Handle errors in check_purchase.
-#  Report failed attempts. Schedule it in the celery beat.
+# @todo #690:30m Schedule check_purchase in the celery beat.
 
 
-@app.task
-def check_purchase():
-    driver = selenium.SiteDriver(site_url=settings.BASE_URL)
-    category_page = selenium.CategoryPage(driver, CategoryPage.objects.first().url)
-    category_page.load()
-    category_page.add_to_cart()
-
-    order_page = selenium.OrderPage(driver)
-    order_page.load()
-    order_page.fill_contacts()
-    order_page.make_order()
+@app.task(
+    bind=True,
+    autoretry_for=(WebDriverException, AssertionError),
+    retry_kwargs={'max_retries': 3},
+)
+def check_purchase(self):
+    try:
+        driver = selenium.SiteDriver(site_url=settings.BASE_URL)
+        category_page = selenium.CategoryPage(driver, CategoryPage.objects.first().url)
+        category_page.load()
+        category_page.add_to_cart()
+
+        order_page = selenium.OrderPage(driver)
+        order_page.load()
+        order_page.fill_contacts()
+        order_page.make_order()
+
+        success_page = selenium.SuccessPage(driver)
+        success_page.load()
+        assert success_page.is_success()
+    except (WebDriverException, AssertionError) as err:
+        if self.request.retries + 1 > self.max_retries:
```

Comment:
one failed retry seems to me enough reason to send the report. Why you choose three ones?
And we should send one report for one failing. With this code celery will spawn our telegram channel

Reply:
>one failed retry seems to me enough reason to send the report

but what if it is just false positive temporary issue, that usually occurs.
just in case i will create a setting variable for retry number

>With this code celery will spawn our telegram channel

It's not, it will only send the message on the last attempt.

Reply:
@duker33 ,
>but this last attempt occurs every max_retries check. It increases an interval of repeated messages, but still called spawn

i don't see the point. Can you describe it in more detail?
I believe it will report an error once at the end of the retry counter.
e.g. if  `max_retries = 3` then the third failed try only will be reported, and previous once will be omitted

>But this problem requires more systematic decision

Let's just set max retries number to 1 and if there are issues with this, we will find a solution

Reply:
> Let's just set max retries number to 1 and if there are issues with this, we will find a solution

@artemiy312 , y, it's perfect.

About `max_retries` - merge this code plz. I'll create issue and we'll discuss it separately